### ISSUE TITLE
Confusion when loading ConfigurationOfSeaside3 (which loads baseline) from ConfiguraitonOfMagrite

### ISSUE THREAD

**Issue by dalehenrich** [dalehenrich/metacello-work#331]
I have BaselineOfSeaside3 locked in an image. I load Magritte3 which references ConfigurationOfSeaside3, but  ConfigurationOfSeaside3 loads BaselineOfSeaside3 and the project registration is pretty confused.

The ConfigurationOfSeaside3 uses a baseline to load Seaside, where is the baselineProjectSpec?

```
.                         -> ConfigurationOfSeaside3 3.1.3.1 from http://smalltalkhub.com/mc/Seaside/MetacelloConfigurations/main
..                        -> anOrderedCollection( BaselineOfGrease [baseline] from github://GsDevKit/Grease:master/repository, BaselineOfRB [baseline] from github://dale...
(class)@                  -> MetacelloProjectRegistration
(oop)@                    -> 246371329
(committed)@              -> true
baselineProjectSpec@      -> nil
configurationProjectSpec@ -> spec name: 'Seaside3'; versionString: '3.1.3.1'; preLoadDoIt: nil; postLoadDoIt: nil; loads: #('Core' 'Javascript' 'RSS' 'Filesystem' ...
loadedInImage@            -> true
locked@                   -> false
mutable@                  -> false
projectName@              -> 'Seaside3'
versionInfo@              -> aMetacelloProjectRegistrationVersionInfo
```

Where's the SHA?

```
.               -> aMetacelloProjectRegistrationVersionInfo
..              -> ConfigurationOfSeaside3 3.1.3.1 from http://smalltalkhub.com/mc/Seaside/MetacelloConfigurations/main
(class)@        -> MetacelloProjectRegistrationVersionInfo
(oop)@          -> 247417089
(committed)@    -> true
projectVersion@ -> 3.1.3.1 [ConfigurationOfSeaside3]
versionString@  -> '3.1.3.1'
```

What the heck is ConfigurationOfMagritte doing in there?

```
.               -> spec name: 'Seaside3'; versionString: '3.1.3.1'; preLoadDoIt: nil; postLoadDoIt: nil; loads: #('Core' 'Javascript' 'RSS' 'Filesystem' ...
..              -> ConfigurationOfSeaside3 3.1.3.1 from http://smalltalkhub.com/mc/Seaside/MetacelloConfigurations/main
(class)@        -> MetacelloMCConfigurationOfProjectSpec
(oop)@          -> 247417345
(committed)@    -> true
className@      -> 'ConfigurationOfSeaside3'
file@           -> nil
loader@         -> nil
loads@          -> anArray( 'Core', 'Javascript', 'RSS', 'Filesystem', 'Welcome')
mutable@        -> false
name@           -> 'Seaside3'
operator@       -> nil
postLoadDoIt@   -> spec value: nil
preLoadDoIt@    -> spec value: nil
project@        -> ConfigurationOfMagritte3(3.0-baseline [ConfigurationOfMagritte3], 3.0 [ConfigurationOfMagritte3], 3.0.1 [ConfigurationOfMagritte3], 3.0.2 [C...
projectPackage@ -> spec name: 'ConfigurationOfSeaside3'; repository: 'http://smalltalkhub.com/mc/Seaside/MetacelloConfigurations/main'.
repositories@   -> spec repository: 'http://smalltalkhub.com/mc/Seaside/MetacelloConfigurations/main'
versionString@  -> '3.1.3.1'
```

**Comment by dalehenrich** [comment]
This is happening because the project ref spec for Seaside in Magritte is named `Seaside3`:

```Smalltalk
baseline310CommonExtDeps: spec
  "Common external dependencies for baseline 3.1.0"

  spec
    project: 'Grease'
      with: [ 
          spec
            className: 'ConfigurationOfGrease';
            versionString: #'stable';
            loads: #('Core');
            repository:
                'http://smalltalkhub.com/mc/Seaside/MetacelloConfigurations/main' ];
    project: 'Seaside3'
      with: [ 
          "note: we do not want to depend on Zinc, since this is not present in Squeak. Currently no adapter is loaded"
          spec
            className: 'ConfigurationOfSeaside3';
            versionString: #'stable';
            loads: #('Core' 'Javascript' 'RSS' 'Filesystem' 'Welcome');
            repository:
                'http://smalltalkhub.com/mc/Seaside/MetacelloConfigurations/main' ]
```

and Metacello is confusing the project reference in Magritte with the actual load spec, since clearly the spec bing merged is from within Magritte:

```
.               -> spec name: 'Seaside3'; versionString: '3.1.3.1'; preLoadDoIt: nil; postLoadDoIt: nil; loads: #('Core' 'Javascript' 'RSS' 'Filesystem' ...
..              -> ConfigurationOfSeaside3 3.1.3.1 from http://smalltalkhub.com/mc/Seaside/MetacelloConfigurations/main
(class)@        -> MetacelloMCConfigurationOfProjectSpec
(oop)@          -> 260824321
className@      -> 'ConfigurationOfSeaside3'
file@           -> nil
loader@         -> nil
loads@          -> anArray( 'Core', 'Javascript', 'RSS', 'Filesystem', 'Welcome')
mutable@        -> nil
name@           -> 'Seaside3'
operator@       -> nil
postLoadDoIt@   -> spec value: nil
preLoadDoIt@    -> spec value: nil
project@        -> ConfigurationOfMagritte3(3.0-baseline [ConfigurationOfMagritte3], 3.0 [ConfigurationOfMagritte3], 3.0.1 [ConfigurationOfMagritte3], 3.0.2 [C...
projectPackage@ -> nil
repositories@   -> spec repository: 'http://smalltalkhub.com/mc/Seaside/MetacelloConfigurations/main'
versionString@  -> '3.1.3.1'
```

**Comment by dalehenrich** [comment]
Okay, this is all happening because in MetacelloProjectSpecForLoad>>performLoad, the project reference spec (whose project is ConfigurationOfMagritte) is converted into a version using the project spec className ... so you end up looking up the version in the correct configuration...

Methinks that MetacelloProjectSpecForLoad>>performLoad should create a proper project spec for the project based on the version ... before:

```Smalltalk
performLoad
  | displayString spec |
  spec := self projectSpec.
  displayString := 'Project: ' , spec name.
  spec versionString ~~ nil
    ifTrue: [ displayString := displayString , ' ' , spec versionString ].
  Transcript
    cr;
    show: displayString.
  self hasOverride
    ifTrue: [ 
      | override |
      override := self overrideProjectSpec copy.
      override mergeScriptLoads: spec.
      override loadVersion: nil ]
    ifFalse: [ 
      spec
        loadVersion:
          (self useDetermineVersionForLoad
            ifTrue: [ spec determineCurrentVersionForLoad ]
            ifFalse: [ spec versionOrNil ]) ]
```

after: 

```Smalltalk
performLoad
  | displayString spec |
  spec := self projectSpec.
  displayString := 'Project: ' , spec name.
  spec versionString ~~ nil
    ifTrue: [ displayString := displayString , ' ' , spec versionString ].
  Transcript
    cr;
    show: displayString.
  self hasOverride
    ifTrue: [ 
      | override |
      override := self overrideProjectSpec copy.
      override mergeScriptLoads: spec.
      override loadVersion: nil ]
    ifFalse: [ | vrsn |
      vrsn := self useDetermineVersionForLoad
            ifTrue: [ spec determineCurrentVersionForLoad ]
            ifFalse: [ spec versionOrNil ].
      vrsn asProjectSpec loadVersion: vrsn ]
```

**Comment by dalehenrich** [comment]
@jbrichau, this bug might be the root cause of you needing to nuke registrations periodically ... I found this while loading Magritte3 which referenced Seaside3 and the registration for Seaside3 wa just hosed ...

**Comment by dalehenrich** [comment]
- [ ] need test case where configuration has to be reloaded to reveal the version ... that's one case where we need to do this right (from a referenced project)

**Comment by dalehenrich** [comment]
Magritte3 and Seaside loaded with correct project registrations ....

**Comment by dalehenrich** [comment]
[clean sheet](https://travis-ci.org/dalehenrich/metacello-work/builds/51450449)

**Comment by jbrichau** [comment]
@dalehenrich your issue sounds a bit familiar indeed. I actually think I got that once with a Zinc project...

**Comment by krono** [comment]
Note there's a test-fix in #348
